### PR TITLE
Multiple parse arguments declarations

### PR DESCRIPTION
**Pull request by alxarsenault**

### CODE REVIEW

**Comment by p-groarke** [comment]
Gonna need a rebase (conflicts)

**Comment by p-groarke** [comment]
Oh and push with your email set in your git, that should add you to contributors I believe.
```
$ git config --global user.name "John Doe"
$ git config --global user.email "john@doe.org"
$ /* do your rebase */
```

Then change the last commit's author and email.
```
$ git commit --amend --author="John Doe <john@doe.org>"
```

File: include/ns_getopt.h [hunk 1]
```diff
@@ -238,10 +263,10 @@ inline void print_help(const std::vector<argument>& args, const char* arg0
 	bool arguments_are_optional = has_flag(option.flags
 			, flag::arguments_are_optional);
 	std::string raw_args = "";
-	bool first = arguments_are_optional;
-	for (const auto& x : args) {
-		if (x.arg_type == type::raw_arg) {
-			raw_args += (first ? " [" : " ") + x.long_arg;
+	bool first = !arguments_are_optional;
```
Comment (p-groarke) [inline include/ns_getopt.h]: shouldn't be negated anymore

File: include/ns_getopt.h [hunk 2]
```diff
@@ -342,26 +367,38 @@ inline void print_help(const std::vector<argument>& args, const char* arg0
...
+		if (x->arg_type == type::raw_arg)
+			x->raw_arg_pos = raw_args_count++;
 	}
 
 	for (int i = 0; i < argc; ++i) {
 		/* First argument is a special snowflake. */
 		if (i == 0 && !has_flag(option.flags, flag::arg0_is_normal_argument)) {
-			if (argc == 1 && !has_flag(option.flags, arguments_are_optional)) {
-				return do_exit(args, option, argv[0]);
+			if (argc == 1 && has_flag(option.flags, arguments_are_optional)) {
```
Comment (p-groarke) [inline include/ns_getopt.h]: Should be negated.

File: include/ns_getopt.h [hunk 3]
```diff
@@ -582,11 +619,11 @@ void print_description(const std::string& s, size_t indentation) {
...
 
-inline bool do_exit(std::vector<argument>& args, const options& option
-		, const char* arg0) {
-	if (!has_flag(option.flags, flag::dont_print_help)) {
-		print_help(args, arg0, option);
-	}
+inline bool do_exit(const argument* args, std::size_t args_size
+		, const options& option , const char* arg0) {
+	print_help(args, args_size, arg0, option);
+	if (has_flag(option.flags, flag::dont_print_help))
```
Comment (p-groarke) [inline include/ns_getopt.h]: This is broken. Copy paste from master.

**Review by p-groarke — CHANGES_REQUESTED** [review]

File: include/ns_getopt.h (hunk 3 above)
Comment (alxarsenault) [reply include/ns_getopt.h]: Done.

File: include/ns_getopt.h (hunk 2 above)
Comment (alxarsenault) [reply include/ns_getopt.h]: Done.

File: include/ns_getopt.h (hunk 1 above)
Comment (alxarsenault) [reply include/ns_getopt.h]: Done.

**Review by p-groarke — APPROVED** [review]

**Comment by p-groarke** [comment]
Awesome!

**Pull request merged**